app/hooks/search.tsx

```diff
@@ -0,0 +1,171 @@
+        const prefixMatch = prefixRe.exec(questionNormalized)
+        const currPosition = fullMatch?.index ?? prefixMatch?.index ?? prevPosition
+        const distanceMultiplier =
+          questionNormalized.slice(prevPosition, currPosition).split(' ').length === 2 ? 1 : 0.9
+
+        if (fullMatch) {
+          score += weight * distanceMultiplier
+        } else {
+          if (prefixMatch) {
+            score += 0.9 * weight * distanceMultiplier
+          } else {
+            score -= 0.2 * weight
+          }
+        }
+        prevPosition = currPosition
+      }
+
+      return score / totalWeight
+    }
+
+    return questions.current
+      .map(({pageid, title}) => {
+        const normalized = normalize(title)
+        return {
+          pageid,
+          title,
+          normalized,
+          model: 'plaintext',
+          score: scoringFn(normalized),
+        }
+      })
+      .sort(byScore)
+      .slice(0, numResults)
+      .filter(({score}) => score > 0)
+  }
+
+/**
+ * Ignore unimportant details for similarity comparison
+ */
+const normalize = (question: string) =>
+  question
+    .toLowerCase()
+    .replace(/[^\w ]|\b(?:an?|the?)\b/g, '')
+    .replace(/(\w{2})s\b/g, '$1') // cannot use lookbehind (?<=...) because not supported on Safari
+    .replace(/\s+|_|&\s*/g, ' ')
+    .trim()
+
+/**
+ * Construct a semantic search function, that will search over the provided questions
+ */
+const semanticSearcher = async (questions: Question[], encodings: Tensor2D, numResults: number) => {
+  const model = await load()
+
+  return async (userQuery: string): Promise<SearchResult[]> => {
+    const searchQuery = userQuery.toLowerCase().trim().replace(/\s+/g, ' ')
+
+    // encodings is 2D tensor of 512-dims embeddings for each sentence
+    const encoding = await model.embed(searchQuery)
+    // numerator of cosine similar is dot prod since vectors normalized
+    const scores = tf.matMul(encoding as unknown as TensorLike, encodings, false, true).dataSync()
+
+    // tensorflow requires explicit memory management to avoid memory leaks
+    encoding.dispose()
+
+    const seen = new Set()
+    return questions
+      .map((question, index) => ({
+        score: scores[index],
+        model: 'tensorflow',
+        ...question,
+      }))
+      .sort(byScore)
+      .filter(({pageid}) => !seen.has(pageid) && seen.add(pageid))
+      .slice(0, numResults)
+  }
+}
+/**
+ * Configure the search engine.
+ *
+ * This will download the embeddings for semantic search, but until that is set up, will
+ * use baseline search over the list of questions already loaded on the site.
+ * Searches containing only one or two words will also use the baseline search
+ */
+export const useSearch = (onSiteQuestions: MutableRefObject<Question[]>, numResults = 5) => {
+  const baselineSearch = makeBaselineSearcher(onSiteQuestions, numResults)
+  const [searchFn, setSearchFn] = useState(() => baselineSearch)
```

Comment:
why calling `makeBaselineSearcher()` every render and then only use it once?

Reply:
its not only used once. It's used every time someone searches for something with 1 or 2 words - https://github.com/StampyAI/stampy-ui/pull/260/files#diff-c7d7c7985c73cdd1f606a88adf3be2bf2c177d2045a2b22523ab502bf6d59218R162

Reply:
the "lazy init" function in `useState(() => baselineSearch)` is only called on mount by React